pkg/lib/oauth/handler/handler_token.go

```diff
@@ -242,20 +242,23 @@ func (h *TokenHandler) validateRequest(r protocol.TokenRequest, client *config.O
 
 var errInvalidAuthzCode = protocol.NewError("invalid_grant", "invalid authorization code")
 
-func (h *TokenHandler) validateApp2AppTokenChallenge(jwt string) (*app2app.Request, error) {
+func (h *TokenHandler) parseApp2AppTokenRequest(jwt string, verifyChallenge bool) (*app2app.Request, error) {
```

Comment:
Instead of introducing a new parameter, it is better to define more methods

```golang
// Note that I changed the method names to start with the prefix app2app. I hope it can make us easier to relate these methods are highly related.
func (h *TokenHandler) app2appParseNoVerify(jwt string) (*app2app.Request, error)
func (h *TokenHandler) app2appVerifyAndConsumeChallenge(r *app2app.Request) error
func (h *TokenHandler) app2appGetDeviceKeyJWK(r *app2app.Request) (jwk.Key, error)
```

So once we have parsed the JWT and got a app2app.Request, we can keep it and use it anyway we want to.

Reply:
It will be complicated, because `updateApp2AppDeviceKeyIfNeeded` will be used with `verifyChallenge = true` and `verifyChallenge = false`.
However only the verify logics are different, others logics should be all same. We can extract many functions to handle it, i.e.
```
app2appGetDeviceKeyJWKUnverified
app2appGetDeviceKeyJWKVerified
app2appUpdateDeviceKeyIfNeededVerified
app2appUpdateDeviceKeyIfNeededUnverified
```

But it seems only complicating the code so I prefer adding a parameter.

Reply:
I tried to restructure the code so that we don't need the unverified cases, please check :adore: